### PR TITLE
Exclude ARM Binaries from BA2025 analysis

### PR DESCRIPTION
**Pull request by marmegh**
Addressing issue #627 by:

- Exclude ARM binaries from BA2025 analysis
- Added ARM not applicable test case (.exe and .pdb files)
- Extended test functionality to confirm correct skip reason is present. This allows confirmation for the current BA2025 test cases and can be used to add granularity to existing/future test cases.

### CODE REVIEW

File: src/BinSkim.Rules/PERules/BA2025.EnableShadowStack.cs [hunk 1]
```diff
@@ -66,6 +66,12 @@ public override AnalysisApplicability CanAnalyzePE(PEBinary target, Sarif.Proper
                 return notApplicable;
             }
 
+            if (portableExecutable.Machine == Machine.Arm || portableExecutable.Machine == Machine.ArmThumb2)
+            {
+                reasonForNotAnalyzing = MetadataConditions.ImageIsArmBinary;
+                return notApplicable;
+            }
```
Comment (marmegh) [inline src/BinSkim.Rules/PERules/BA2025.EnableShadowStack.cs:73]: This is the only logic update for this change. Specifically, excluding ARM binaries from BA2025 analysis in a similar fashion as ARM64 binaries.

File: src/BinSkim.Sdk/MetadataConditions.cs [hunk 2]
```diff
@@ -10,6 +10,7 @@ public static class MetadataConditions
         public static readonly string ImageIsNotExe = SdkResources.MetadataCondition_ImageIsNotExe;
         public static readonly string ImageIsNotMachO = SdkResources.MetadataCondition_ImageIsNotMachO;
         public static readonly string CouldNotLoadPdb = SdkResources.MetadataCondition_CouldNotLoadPdb;
+        public static readonly string ImageIsArmBinary = SdkResources.MetadataCondition_ImageIsArmBinary;
```
Comment (marmegh) [inline src/BinSkim.Sdk/MetadataConditions.cs:13]: >public static readonly string ImageIsArmBinary = SdkResources.MetadataCondition_ImageIsArmBinary; [](http://example.com/codeflow?start=8&length=97)

Add a MetadataConditions entry for ARMBinaries. This is separate as changing the previous user facing string could confuse customers by changing details in the results.

File: src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs [hunk 3]
```diff
@@ -276,6 +298,26 @@ private BinaryAnalyzerContext CreateContext(TestMessageLogger logger, Properties
...
+        }
+
+        private void VerifyApplicabilityResults(
+            BinarySkimmer skimmer,
+            HashSet<string> targets,
+            bool useDefaultPolicy,
+            AnalysisApplicability expectedApplicability,
+            string ruleName,
+            string expectedReasonForNotAnalyzing)
+        {
```
Comment (marmegh) [inline src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs:317]: Created method for shared code.

File: src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs [hunk 4]
```diff
@@ -460,6 +498,17 @@ private HashSet<string> GetTestFilesMatchingConditions(HashSet<string> metadataC
                 result.Add(Path.Combine(testFilesDirectory, "DotnetNative_x86_VS2019_UniversalApp.exe"));
             }
 
+            if (metadataConditions.Contains(MetadataConditions.ImageIsArmBinary))
+            {
+                result.Add(Path.Combine(testFilesDirectory, "ARM_CETShadowStack_NotApplicable.exe"));
```
Comment (shaopeng-gh) [inline src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs:503]: >ARM_CETShadowStack_NotApplicable.exe [](http://example.com/codeflow?start=61&length=36)

nice, how did you generate this file, please add the step in md file 
\docs\FunctionalTestBuildScripts.md

File: src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs [hunk 5]
```diff
@@ -276,6 +294,26 @@ private BinaryAnalyzerContext CreateContext(TestMessageLogger logger, Properties
...
+        private void VerifyApplicabilityResults(
+            BinarySkimmer skimmer,
+            HashSet<string> targets,
+            bool useDefaultPolicy,
+            AnalysisApplicability expectedApplicability,
+            string ruleName,
+            string expectedReasonForNotAnalyzing)
+        {
+
+            var context = new BinaryAnalyzerContext();
```
Comment (shaopeng-gh) [inline src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs:315]: first line empty can be removed.

**Review by shaopeng-gh — CHANGES_REQUESTED** [review]
:clock1:

File: src/ReleaseHistory.md [hunk 6]
```diff
@@ -7,6 +7,7 @@
 * Bump ELFSharp from 2.14.0 to 2.15.0. [#631](https://github.com/microsoft/binskim/pull/631)
 * FEATURE: Enable BinSkim for MacOS. [#576](https://github.com/microsoft/binskim/pull/576)
 * Bump Sarif.Sdk by updating submodule from [4e9f606 to fc9a9df](https://github.com/microsoft/sarif-sdk/compare/4e9f606bb0e88428866e253352cdc70dc68f98cb...fc9a9dfb865096b5aaa9fa3651854670940f7459). [#638](https://github.com/microsoft/binskim/pull/638)
+* BUGFIX: Fix `BA2025.EnableShadowStack` warning for ARM Binaries which cannot use `/CETCOMPAT`. [#650](https://github.com/microsoft/binskim/pull/650)
```
Comment (shaopeng-gh) [inline src/ReleaseHistory.md:10]: >BUGFIX: Fix `BA2025.EnableShadowStack` warning for ARM Binaries which cannot use `/CETCOMPAT`. [](http://example.com/codeflow?start=2&length=94)

I see a sample below:
FALSE POSITIVE FIX: Skip PDB-driven analysis for the generated .NET core native bootstrap exe (which is not user-controllable code).
See if we think this template is better for this change 
FALSE POSITIVE FIX: Skip rule xxx for xxx (reason)

File: docs/FunctionalTestBuildScripts.md [hunk 7]
```diff
@@ -47,3 +47,40 @@ Also create two user functions `userfn_use_safebuffers_1()` and `userfn_use_safe
 A simple `Windows Kernel Mode Driver` program, created with `Visual Studio 2019` that generates a .exe and associated .pdb file. Code to reproduce:  
 Use `NTSTATUS GsDriverEntry(_In_ PDRIVER_OBJECT DriverObject, _In_ PUNICODE_STRING RegistryPath)` as entry point and do not decorated with `__declspec(safebuffers)`.
 No user functions decorated with `__declspec(safebuffers)`.
+
+## ARM64_CETShadowStack_NotApplicable.exe
```
Comment (shaopeng-gh) [inline docs/FunctionalTestBuildScripts.md:51]: add a empty line consistent with others

**Review by shaopeng-gh — APPROVED** [review]
:shipit:

File: src/BinSkim.Rules/PERules/BA2025.EnableShadowStack.cs (hunk 1 above)
Comment (michaelcfanning) [reply src/BinSkim.Rules/PERules/BA2025.EnableShadowStack.cs:73]: Nice code comment! Absolutely the info I was looking for. :)

File: src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs [hunk 8]
```diff
@@ -1166,10 +1214,19 @@ public void BA2025_EnableShadowStack_Fail()
         [Fact]
         public void BA2025_EnableShadowStack_NotApplicable()
         {
-            this.VerifyApplicability(
+            HashSet<string> notApplicableArm64 = new HashSet<string>() { MetadataConditions.ImageIsArm64BitBinary };
+
+            this.VerifyApplicabililtyByConditionsOnly(
```
Comment (michaelcfanning) [inline src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs:1219]: >VerifyApplicabililtyByConditionsOnly [](http://example.com/codeflow?start=17&length=36)

You are getting solidly into our test infrastructure if you understand this pattern. nice work. :) I'm sure Shaopeng may have been helpful in gaining 'the knowledge'.

**Review by michaelcfanning — APPROVED** [review]
:shipit:

**Pull request merged**